src/leaflike/bookmarks.clj

```diff
@@ -16,98 +16,130 @@
       (when (not-empty tags)
         (tags-db/create tags)
         (bm-db/tag-bookmark bookmark-id tags))
       (-> (res/redirect "/bookmarks")
           (assoc-in [:headers "Content-Type"] "text/html")))
     (assoc (res/redirect (str "/bookmarks/edit/" (:id params)))
-           :flash {:error-msg (bm-spec/describe-errors params)})))
+      :flash {:error-msg (bm-spec/describe-errors params)})))
+
+(defn authorize-edit
+  [{:keys [params] :as request}]
+  (let [username (get-in request [:session :username])
+        user-id (:id (user-db/get-user-if-exists username))
+        created_by (:created_by (into {} (-> (Integer/parseInt (:id params))
+                                             (bm-db/get-created-by))))]
+    (if (= user-id created_by)
+      (edit request)
+      (assoc (res/redirect (str "/bookmarks/edit/" (:id params)))
+        :flash {:error-msg (str "Only " (:username (into {} (user-db/get-username-by-user-id created_by)))
+                                " can edit this bookmark")}))))
 
 (def items-per-page 10)
 
 (defn fetch-bookmarks
   [username {:keys [tag search-terms] :as params}]
   (let [user (hutils/get-user {:session {:username username}})
         page (dec (:page params))
-        query {:user-id (:id user)
-               :tag tag
+        query {:user-id      (:id user)
+               :tag          tag
                :search-terms search-terms}
         bookmarks (bm-db/fetch-bookmarks (merge query
-                                                {:limit items-per-page
+                                                {:limit  items-per-page
                                                  :offset (* items-per-page page)}))
         num-bookmarks (bm-db/count-bookmarks query)]
     {:bookmarks bookmarks
      :num-pages (max 1
                      (int (Math/ceil (/ num-bookmarks items-per-page))))}))
 
-(defn list-by-id
+(defn list-by-creator-id
   [{:keys [route-params] :as request}]
-  (let [id        (:id route-params)
-        user      (hutils/get-user request)
-        params    {:id id :user-id (:id user)}]
+  (let [id (:id route-params)
+        user (hutils/get-user request)
+        params {:id id :user-id (:id user)}]
     (if (s/valid? :leaflike.bookmarks.spec/id id)
       (bm-db/list-by-id params)
       (assoc (res/redirect "/bookmarks")
-             :flash {:error-msg "Invalid bookmark id"}))))
+        :flash {:error-msg "Invalid bookmark id"}))))
 
 (defn delete
   [{:keys [params] :as request}]
-  (let [user      (hutils/get-user request)
+  (let [user (hutils/get-user request)
         unparsed-id (:id params)]
     (if (s/valid? :leaflike.bookmarks.spec/id unparsed-id)
       (let [id (Integer/parseInt unparsed-id)]
-        (do (bm-db/remove-all-tags id)
-            (bm-db/delete {:id id
-                           :user-id (:id user)})
+        (if (= (:created_by (reduce conj {} (bm-db/get-created-by id))) (:id user))
```

Comment:
why is reduce needed here? 

`(:created_by (reduce conj {} [{:created_by 1}]))` => `(:created_by (first [{:created_by 1}]))`

Reply:
I think here you should think about what do you expect from your db interface and make the fn return the expected value. like reading the fn name `get-created-by`, I expect to get the id of the user that created the bookmark which is your function is not returning. you are returning a list of query results and expecting your domain fns to interpret that result. the data structures from your querying interface are bleeding into your domain logic. also maybe name the fn better, `get-created-by` is a consequence of the field that exists in the db. I would name this fn `bookmark/get-creator` (even namespace aliasing should be reader-friendly).

Reply:
> why is reduce needed here?
> 
> `(:created_by (reduce conj {} [{:created_by 1}]))` => `(:created_by (first [{:created_by 1}]))`

I think it was because `(:created_by (first [{:created_by 1}]))` returns a vector, but I wanted a map.

Reply:
> I think here you should think about what do you expect from your db interface and make the fn return the expected value. like reading the fn name `get-created-by`, I expect to get the id of the user that created the bookmark which is your function is not returning. you are returning a list of query results and expecting your domain fns to interpret that result. the data structures from your querying interface are bleeding into your domain logic. also maybe name the fn better, `get-created-by` is a consequence of the field that exists in the db. I would name this fn `bookmark/get-creator` (even namespace aliasing should be reader-friendly).

`get-created-by` returns a sequence of a map like so: `({:created_by 1})`, isn't that what I'd expect get-created-by to return? The sequence always consists of just one map. I think the problem here is that I leave it up to the consumer of this function to operate on that sequence of map to get the value of created_by. Ideally, I would like this function to return a map (or maybe just the id) instead of a sequence of that map -- and I hope to work on that. Let me know if we're on the same page here. If you're addressing a different problem, we can talk about it in person.

Also, yeah I totally agree - `get-created-by` is a weird name. Not sure what I was thinking. I think I initially did call it `get-creator` and then changed it to this. I'll ensure that I'm not always trying to match the column name.

Reply:
Yup, I did call it `get-creator` at one point https://github.com/nilenso/leaflike/blob/bfdb5c04cce2e9892485ac5ac3e63a00e4d16cfa/src/leaflike/bookmarks/db.clj#L203